src/libraries/System.Text.Json/tests/Common/JsonSchemaExporterTests.TestTypes.cs

```diff
@@ -0,0 +1,1320 @@
+                                }
+                            }
+                        }
+                    }
+                """,
+                Options: new JsonSchemaExporterOptions
+                {
+                    OnSchemaNodeGenerated = static (ctx, schema) =>
+                    {
+                        if (ctx.TypeInfo.Kind is JsonTypeInfoKind.None)
+                        {
+                            return;
+                        }
+
+                        string idPath = ctx.PropertyInfo is { } property
+                            ? ctx.TypeInfo.Type.Name.ToLower() + "/" + property.Name.ToLower()
+                            : ctx.TypeInfo.Type.Name.ToLower();
+
+                        string idUrl = "https://example.com/schema/" + idPath + ".json";
+
+                        if (schema.ContainsKey("$ref"))
+                        {
+                            schema["$ref"] = idUrl;
+                        }
+                        else
+                        {
+                            schema["$id"] = idUrl;
+                        }
+                    }
+                });
+
+            yield return new TestData<PocoWithRecursiveCollectionElement>(
+                Value: new() { Children = [new(), new() { Children = [] }] },
+                ExpectedJsonSchema: """
+                {
+                    "type": ["object","null"],
+                    "properties": {
+                        "Children": {
+                            "type": "array",
+                            "items": { "$ref" : "#" }
+                        }
+                    }
+                }
+                """);
+
+            yield return new TestData<PocoWithRecursiveDictionaryValue>(
+                Value: new() { Children = new() { ["key1"] = new(), ["key2"] = new() { Children = new() { ["key3"] = new() }  } } },
+                ExpectedJsonSchema: """
+                {
+                    "type": ["object","null"],
+                    "properties": {
+                        "Children": {
+                            "type": "object",
+                            "additionalProperties": { "$ref" : "#" }
+                        }
+                    }
+                }
+                """);
+
+            yield return new TestData<PocoWithDescription>(
+                Value: new() { X = 42 },
+                ExpectedJsonSchema: """
+                {
+                  "description": "The type description",
+                  "type": ["object","null"],
+                  "properties": {
+                    "X": {
+                      "description": "The property description",
+                      "type": "integer"
+                    }
+                  }
+                }
+                """,
+                Options: new()
+                {
+                    OnSchemaNodeGenerated = static (ctx, schema) =>
```

Comment:
Not at this stage, no. STJ doesn't even reference `System.ComponentModel` so any such handlers need to be defined further down the stack.

Reply:
No, I mean, for example, your handler to add `$anchor`.  Would you be making common use cases like that available?  (This has been done in the past for things like the JSON encoder on the serializer options.)

Reply:
If we do that we should consider changing the handler type to be a single-method interface (as with `IJsonTypeInfoResolver`) so that nominal types can implement it. Needs second round of API review though before we can proceed.